### DISCUSSION TITLE
How to create a Seurat Object from 10x Visium data if the feature-barcodes dataset is not in hdf5 file format?

### DISCUSSION THREAD

**Discussion by ssingh95jhu** [satijalab/seurat#8926] [Q&A]
I am trying to create a Seurat Object from 10x Visium data. I have the following files for the tissue of interest:
1. matrix.mtx.gz
2. barcodes.tsv.gz
3. features.tsv.gz
4. and a folder called "spatial" containing the relevant files: scale factors_json.json, tissue_hires_image.png, tissue_positions.csv

I am following the steps provided here: [https://satijalab.org/seurat/reference/load10x_spatial](url)

However, I do not have my 10x Visium feature-barcodes matrix (data) in the hdf5 format (i.e., filtered_feature_bc_matrix.h5)

The strategy which I followed was:
1. First, created a Seurat object using the Read10X function using the matrix.mtx.gz, barcodes.tsv.gz and features.tsv.gz files.
2. Then, I tried to add the images to the above Seurat object but was not successful. This prevents me from implementing functions like SpatialFeaturePlot or SpatialDimplot.

Is there a work around for this? * I came across similar discussion threads here but couldn't find a solution *
I would really appreciate if you have any suggestions.

Regards

**Comment by lvmt** [comment]
hi
i meet the same question and  i had solved it.  

the error is because  the 'tissue_lowres_image.png' lost  

the Solution：
1 `cp tissue_hires_image.png tissue_lowres_image.png`
2 modify `scalefactors_json.json` file,  let `tissue_lowres_scalef.value = tissue_hires_scalef.value ` 
```
matrix_dir = 'filtered_feature_bc_matrix/' 
counts = Seurat::Read10X(data.dir = matrix_dir)  

data = Seurat::CreateSeuratObject(
    counts = counts , 
    project = 'test', 
    assay = 'Spatial')

data$slice = 1 
data$region = 'test' 

imgpath = "spatial"
img = Seurat::Read10X_Image(image.dir = imgpath)  

Seurat::DefaultAssay(object = img) <- 'Spatial'  

img = img[colnames(x = data)]  
data[['image']] = img  

SpatialFeaturePlot(data, features = "nCount_Spatial")

```

hope this can help U.

**Reply by ssingh95jhu** [reply]
Thanks lvmt for your suggestion! I took your advice and made some changes to my workflow and achieved partial success.

1. Found out the "tissue_lowres_image.png" file and pasted it back in the "spatial" folder already containing the "tissue_hires_image.png" file. Never made any changes to the "scalefactors_json.json" file as you had suggested.

2. Implemented the code as provided by you. SpatialFeaturePlot function seems to be working when I pass the features argument, features="nCount_Spatial". (Image attached: It's a human kidney biopsy tissue)
![Screenshot 2024-05-29 at 3 12 04 PM](https://github.com/satijalab/seurat/assets/141043853/dfbf0ecf-33af-4023-a9e8-6affaf7a5b04)

3. But when I change it to a gene name, for example: features = "LCN2", I get an error as follows:
Warning: No layers found matching search pattern provided
Error in FetchData.Assay5(object = object[[DefaultAssay(object = object)]],  : 
  layer "data" is not found in the object

Have I made any mistake or missing anything here? Once again thanks so much for your help!

**Reply by ccoate** [reply]
setwd("/path/to/folder/ST_data")

# In your folder make a subfolder with barcodes, features, and matrix (.gz for all), named  filtered_feature_bc_matrix

filter_matrix <- Read10X("./filtered_feature_bc_matrix/")

write10xCounts("./filtered_feature_bc_matrix.h5", filter_matrix, type = "HDF5",
               genome = "grch38", version = "3", overwrite = TRUE,
               gene.id = rownames(filter_matrix),
               gene.symbol = rownames(filter_matrix))

This will output a Seurat-compatible filtered_feature_bc_matrix.h5 in your ST_data folder

**Comment by lvmt** [comment]
hi, 

you had successed. 
the error 'layer "data" is not found in the object' is abot data processing。  

because  the created seurat obj just contain `count layer` 


follow this tutorial `https://satijalab.org/seurat/articles/pbmc3k_tutorial` and start from `Standard pre-processing workflow` part

hope this can help U.
